keps/sig-instrumentation/draft-20190404-kubernetes-control-plane-metrics-stability.md

```diff
@@ -0,0 +1,248 @@
+
+### Goals
+
+ * Describe the various stability guarantees for control-plane metrics.
+ * Define a uniform mechanism for expressing metric stability.
+
+### Non-Goals
+
+* We are __*not*__ defining which specific metrics are actually stable.
+
+## Background
+
+Kubernetes control-plane binaries (i.e. scheduler, kubelet, controller-manager, apiserver) use a prometheus client to export binary-specific metrics to a ‘/metrics’ endpoint in prometheus format. Metrics are first defined and then instantiated; later they are registered to a metrics registry. The http handler for the metrics endpoint then delegates responses to the underlying registry.
+
+For the remainder of this document, I will refer to the following terms by these definitions:
+
+* __metric definition__ - this refers to defining a metric. In kubernetes, we use the standard prometheus pattern of using an options struct to define name, type, description of a metric.
+* __metric instantiation__ - this refers to creating an instance of a metric. A metric definition is passed into a metric constructor which, in kubernetes, is a prometheus metric constructor (example).
+* __metric enrollment__ - after being defined and created, individual metrics are officially enrolled to a metrics registry (currently a global one).
+* __metric registration process__ - I use this to refer to the entire lifecycle of a metric from definition, to instantiation, then enrollment.
+
+The fact that the metric registration process always involves these steps is significant because it allows for the possibility of injecting custom behavior in and around these steps.
+
+## Proposal
+
+This KEP proposes a programmatic mechanism to express the stability of a given control plane metric. Individual metrics would be quasi-versioned, i.e. they would have additional bits of metadata which would indicate whether that metric was alpha (not-stable), stable, or deprecated. Metric stability guarantees would depend on the values of those additional bits.
+
+Specifically, this would involve injecting custom behavior during metric registration process by wrapping metric definition, instantiation and enrollment.
+
+### Metric Definition Phase
+
+Currently, the metric definition phase looks like this:
+
+```go
+var someMetricDefinition = prometheus.CounterOpts{
+    Name: "some_metric",
+    Help: "some description",
+}
+```
+
+Since we are using the prometheus provided struct, we are constrained to prometheus provided fields. However, using a custom struct affords us the following:
+```go
+var deprecatedMetricDefinition = kubemetrics.CounterOpts{
+    Name: "some_deprecated_metric",
+    Help: "some description",
+    DeprecatedVersion: "1.15", // this is a custom metadata field
+}
+
+var alphaMetricDefinition = kubemetrics.CounterOpts{
+    Name: "some_alpha_metric",
+    Help: "some description",
+    IsAlpha: true, // this is also a custom metadata field
```

Comment:
perhaps this should be `IsStable` so that all metrics default to being alpha and we avoid accidental promotions to stable?

Reply:
What do you think about adding an enum and making each metric definition explicitly declare the stability class? That way, if we decide to add 'beta' in the future, the API will be nicer.

Reply:
I actually do like that, and had that thought at one point, but it slipped my mind while context-switching :-)

Reply:
It seems we are plagued by the same disease. I also considered this earlier and then forgot about it. Thanks for bringing it up :)

Reply:
I swapped to the enum as discussed. However, I still opted to make it optional, since if DeprecatedVersion can only apply to a stable metric, so having both felt redundant. 

I don't really have a strong opinion on whether it should be optional for deprecated metrics; if you prefer the explicit approach I don't mind making the change.

Reply:
Nevermind, I have made it such that the stability level is always explicit.